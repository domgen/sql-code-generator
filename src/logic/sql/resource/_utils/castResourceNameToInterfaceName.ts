import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
