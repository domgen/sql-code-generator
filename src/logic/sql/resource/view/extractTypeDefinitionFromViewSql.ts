import { extractTypeDefinitionFromQuerySql } from '../../query/extractTypeDefinitionFromQuerySql';
import { TypeDefinitionOfResourceView } from '../../../../model/valueObjects/TypeDefinitionOfResourceView';
/*
  note: a view is effectively an alias for a query
    - with the condition that the query can not have input variables

  therefore, the types are a subset of the query types
*/
export const extractTypeDefinitionFromViewSql = ({ name, sql }: { name: string; sql: string }) => {
  // 0. grab the query definition from the view; e.g.: `CREATE VIEW __NAME__ AS SELECT ....` => `SELECT ...` => typedef
  const querySql = sql
    .split(/(SELECT|select)/)
    .slice(1) // slice(1) since part[0] = CREATE VIEW ... AS, part[1] = SELECT, part[2] = __SELECT_BODY__
    .join(''); // merge to have SELECT __SELECT_BODY__
  const queryDef = extractTypeDefinitionFromQuerySql({ sql: querySql });

  // 1. check that query def does not have any inputs, as views with inputs in query are invalid
  if (false) throw new Error(`query def for view '${name}' can not have inputs`); // TODO: when inputs are defined on query

  // 2. grab the selectExpressions and tableReferences off of the query
  return new TypeDefinitionOfResourceView({
    name,
    selectExpressions: queryDef.selectExpressions,
    tableReferences: queryDef.tableReferences,
  });
};
