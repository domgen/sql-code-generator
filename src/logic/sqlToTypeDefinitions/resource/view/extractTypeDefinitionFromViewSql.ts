import { TypeDefinitionOfResourceView } from '../../../../domain/objects/TypeDefinitionOfResourceView';
import { extractTypeDefinitionFromQuerySql } from '../../query/extractTypeDefinitionFromQuerySql';

/*
  note: a view is effectively a named alias for a query
    - with the condition that the query can not have input variables

  therefore, the types are a subset of the query types
*/
export const extractTypeDefinitionFromViewSql = ({
  name,
  sql,
}: {
  name: string;
  sql: string;
}) => {
  // 0. grab the query definition from the view; e.g.: `CREATE VIEW __NAME__ AS SELECT ....` => `SELECT ...` => typedef
  const querySql = sql
    .split(/(SELECT|select)/)
    .slice(1) // slice(1) since part[0] = CREATE VIEW ... AS, part[1] = SELECT, part[2] = __SELECT_BODY__
    .join(''); // merge to have SELECT __SELECT_BODY__
  const queryDef = extractTypeDefinitionFromQuerySql({
    name,
    path: 'none', // we dont currently track paths for view - and it wont be a problem downstream to fake it, so put a placeholder value
    sql: querySql,
  });

  // 1. check that query def does not have any inputs, as views with inputs in query are invalid
  if (queryDef.inputVariables.length)
    throw new Error(`query def for view '${name}' can not have inputs`);

  // 2. grab the selectExpressions and tableReferences off of the query
  return new TypeDefinitionOfResourceView({
    name,
    selectExpressions: queryDef.selectExpressions,
    tableReferences: queryDef.tableReferences,
  });
};
