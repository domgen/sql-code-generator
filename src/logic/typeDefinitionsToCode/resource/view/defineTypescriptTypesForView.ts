import { ResourceType, TypeDefinition } from '../../../../model';
import { castResourceNameToTypescriptTypeName } from '../../common/castResourceNameToTypescriptTypeName';
import { TypeDefinitionOfResourceView } from '../../../../model/valueObjects/TypeDefinitionOfResourceView';
import { defineTypescriptTypeFromReference } from '../../common/defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypesForView = ({
  definition,
  allDefinitions,
}: {
  definition: TypeDefinitionOfResourceView;
  allDefinitions: TypeDefinition[];
}) => {
  // define column types in typescript format
  const typescriptInterfaceColumnDefinitions = definition.selectExpressions.map((selectExpression) => {
    const typescriptTypeForReference = defineTypescriptTypeFromReference({
      reference: selectExpression.typeReference,
      queryTableReferences: definition.tableReferences,
      typeDefinitions: allDefinitions,
    });
    return `${selectExpression.alias}: ${typescriptTypeForReference};`;
  });

  // output
  const typescriptInterfaceDefinition = `
export interface ${castResourceNameToTypescriptTypeName({ name: definition.name, resourceType: ResourceType.VIEW })} {
  ${typescriptInterfaceColumnDefinitions.join('\n  ')}
}
  `.trim();

  // return typescript types
  return typescriptInterfaceDefinition;
};
