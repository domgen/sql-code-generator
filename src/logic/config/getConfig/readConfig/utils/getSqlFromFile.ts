import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
