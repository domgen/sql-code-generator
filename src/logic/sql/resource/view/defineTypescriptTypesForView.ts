import { ResourceType } from '../../../../model';
import { castResourceNameToInterfaceName } from '../_utils/castResourceNameToInterfaceName';
import { TypeDefinitionOfResourceView } from '../../../../model/valueObjects/TypeDefinitionOfResourceView';

export const defineTypescriptTypesForView = ({ definition }: { definition: TypeDefinitionOfResourceView }) => {
  // define column types in typescript format
  const typescriptInterfaceColumnDefinitions = definition.selectExpressions.map((selectExpression) => {
    // grab the source table name and source table column name
    const [sourceTableAlias, sourceTableColumnName] = selectExpression.sourcePath.split('.');
    const sourceTableName = definition.tableReferences.find((ref) => ref.alias === sourceTableAlias)?.tableName;
    if (!sourceTableName) {
      throw new Error(
        `table alias for of select expression "${selectExpression.sourcePath}" not found in view table references`,
      );
    }
    const sourceTableInterfaceName = castResourceNameToInterfaceName({
      name: sourceTableName,
      resourceType: ResourceType.TABLE,
    });
    return `${selectExpression.alias}: ${sourceTableInterfaceName}['${sourceTableColumnName}'];`;
  });

  // output
  const typescriptInterfaceDefinition = `
export interface ${castResourceNameToInterfaceName({ name: definition.name, resourceType: ResourceType.VIEW })} {
  ${typescriptInterfaceColumnDefinitions.join('\n  ')}
}
  `.trim();

  // return typescript types
  return typescriptInterfaceDefinition;
};
