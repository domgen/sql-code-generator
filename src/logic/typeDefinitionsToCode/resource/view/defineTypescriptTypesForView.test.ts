import { defineTypescriptTypesForView } from './defineTypescriptTypesForView';
import { extractTypeDefinitionFromViewSql } from '../../../sqlToTypeDefinitions/resource/view/extractTypeDefinitionFromViewSql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('defineTypescriptTypesForView', () => {
  it('should generate an accurate looking interface for this table definition', async () => {
    const definition = extractTypeDefinitionFromViewSql({
      name: 'image',
      sql: await getSqlFromFile({
        filePath: `${__dirname}/../../../__test_assets__/views/view_suggestion_current.sql`,
      }),
    });
    const code = defineTypescriptTypesForView({ definition });
    expect(code).toMatchSnapshot();
  });
});
