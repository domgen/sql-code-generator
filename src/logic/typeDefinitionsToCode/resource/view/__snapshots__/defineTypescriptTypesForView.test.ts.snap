// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptTypesForView should generate an accurate looking interface for a view joining three tables and selecting from two 1`] = `
"export interface SqlViewImage {
  id: SqlTableSuggestion['id'];
  uuid: SqlTableSuggestion['uuid'];
  suggestion_source: SqlTableSuggestion['suggestion_source'];
  external_id: SqlTableSuggestion['external_id'];
  status: SqlTableSuggestionVersion['status'];
  result: SqlTableSuggestionVersion['result'];
  created_at: SqlTableSuggestion['created_at'];
  effective_at: SqlTableSuggestionVersion['effective_at'];
  updated_at: SqlTableSuggestionVersion['created_at'];
}"
`;
